common-services-api/src/main/java/org/pentaho/big/data/api/cluster/service/locator/impl/NamedClusterServiceLocatorImpl.java

```diff
@@ -54,28 +54,34 @@ public class NamedClusterServiceLocatorImpl implements NamedClusterServiceLocato
   @VisibleForTesting final Map<String, Map<Class<?>, List<NamedClusterServiceFactory<?>>>> serviceVendorTypeMapping;
   private final ReadWriteLock readWriteLock;
   @VisibleForTesting final String internalShim;
-  private final MetastoreLocator metastoreLocator;
+  private MetastoreLocator metastoreLocator;
   private final NamedClusterService namedClusterManager;
 
   private static final Logger logger = LoggerFactory.getLogger( NamedClusterServiceLocatorImpl.class );
 
   public NamedClusterServiceLocatorImpl( String internalShim, NamedClusterService namedClusterManager ) {
-    MetastoreLocator metastoreLocator1;
     this.internalShim = Objects.requireNonNull(
       internalShim, "Set internal.shim in karaf/etc/pentaho.shim.cfg" );
-    try {
-      Collection<MetastoreLocator> metastoreLocators = PluginServiceLoader.loadServices( MetastoreLocator.class );
-      metastoreLocator1 = metastoreLocators.stream().findFirst().get();
-    } catch ( Exception e ) {
-      metastoreLocator1 = null;
-      logger.error( "Error getting metastore locator", e );
-    }
-    metastoreLocator = metastoreLocator1;
     this.namedClusterManager = namedClusterManager;
     readWriteLock = new ReentrantReadWriteLock();
     serviceVendorTypeMapping = new HashMap<>();
   }
 
+  protected MetastoreLocator getMetastoreLocator() {
+    if ( this.metastoreLocator == null ) {
+      MetastoreLocator metastoreLocator1;
+      try {
```

Comment:
Same question as the other diff... this isn't thread safe, does it need to be?

Reply:
No, this shouldn't need to be thread safe.